Create James Waldrop podcast episode post (#5103)
sourcegraph/about: podcast/2.16.md

Click the audio player below to listen, or [click here to watch the video](https://youtu.be/qTwDfW_T7Bg).

<!-- END SUMMARY -->

<!-- START SHOWNOTES -->

JMeter: https://jmeter.apache.org/

Review Board: https://www.reviewboard.org/

Pants: https://www.pantsbuild.org/

Bazel: https://bazel.build/

Gerrit: https://www.gerritcodereview.com/

The Engineering/Management Pendulum: https://charity.wtf/2017/05/11/the-engineer-manager-pendulum/

Bill Gates tweet: https://twitter.com/billgates/status/1206385201594671105?lang=en

FreeCell: https://www.free-freecell-solitaire.com/

James’s Twitter: https://twitter.com/HiveTheory

<!-- END SHOWNOTES -->

<!-- START TRANSCRIPT -->

**Beyang Liu:**

> Hello and welcome to another edition of the Sourcegraph Podcast. My name is Beyang, and on this show, we talk with the people behind amazing dev tools and developer experiences: open source authors, maintainers, dev ex engineers inside forward-thinking companies, developer advocates and educators, founders of dev tool companies–anyone and everyone who brings joy and delight into our lives as programmers. We talk about their work, we talk about their lives, we talk about their vision of what lies ahead for the world of code. Stay tuned.

Reviewer: We typically leave out this intro, but could choose to leave it in.

> All right, so I'm here with James Waldrop today. James, thanks so much for joining. James is the vice president of engineering at Fitbit, now part of Google. His career encompasses many of the great engineering organizations in Silicon Valley, starting out in the 90s where he worked at many different startups in various roles ranging from software development, engineering management, project management, sales support, investor relations, and demo presenter to Bill Gates. More on that in a bit.

Reviewer: Might want to capitalize James' title here.

**Beyang Liu:**

What was the Indy from SGI?

**James Waldrop:**

SGI? Silicon Graphics.

**Beyang Liu:**

Silicon Graphics, yeah.

**James Waldrop:**

So, the Indy was kind of like a personal workstation that had really good 3D computing. A graphics workstation for individuals as opposed to one of those big mainframe systems like an Onix.

**Beyang Liu:**

I see. It sounds very ahead of its time.

**James Waldrop:**

Yeah. Well, we all know what happened with SGI, probably. If not, we can Google it.

Everything they did was ahead of its time. That's pretty much the definition of SGI right there. So, yeah, you configured your DNS settings in a 3D world. If that sounds really silly, well then you know the kind of work we did.

**Beyang Liu:**

> Wow. I've seen stuff where people manage docker containers in Minecraft, but ...

Reviewer: ```suggestion
Wow. I've seen stuff where people manage Docker containers in Minecraft, but ...
```

**James Waldrop:**

Yes, it's a lot like that. Just about that silly too.

**Beyang Liu:**

That as an enterprise product?

**James Waldrop:**

Yeah. Literally you would unbox it, and you needed a great experience. Of course, it needed to be 3D because it was SGI, so they hired us to build it for them.

**Beyang Liu:**

Wow.

**James Waldrop:**

Microsoft, at the time, had a technology that never made it out to the world. Hopefully no lawyers will come after me, but it was called Chrome, which is funny now since I'm in a Chrome browser looking at you. But Chrome was a 3D browser on the desktop. It was like, let's turn all of the normal 2D widgets into 3D. They needed a really cool demo.

> I don't know how many 3D graphics nerds are going to be on your podcast, but it was a technology that originally was at a Sun called TBAG, and the team ended up in Microsoft Research. It was all about what I would call functional 3D graphics. Functional, meaning, in the way that programming languages can be functional. And so, it was a really interesting idea. I still think it was a great idea. It was like a technology in search of a problem, right?

Reviewer: ```suggestion
I don't know how many 3D graphics nerds are going to be on your podcast, but it was a technology that originally was at Sun called TBAG, and the team ended up in Microsoft Research. It was all about what I would call functional 3D graphics. Functional, meaning, in the way that programming languages can be functional. And so, it was a really interesting idea. I still think it was a great idea. It was like a technology in search of a problem, right?
```

And so, you need to hand people tools that really meet them where they are instead of tying too tightly. I think people get trapped by: you have a problem and you end up with the first-order solution to that problem and it's too tightly fit to the problem instead of being fit to the solution.

**Beyang Liu:**

Yeah. There is this difference between dev tools and other software applications that are meant for non-developers. I feel that in the latter case, you're always trying to produce some kind of on-rails experience that doesn't allow for too much diversion from the core workflow because the idea is that you've thought of how to do this and the app guides you through that process. Whereas with developers, it's all about, like, oh, I don't want to do the same thing over and over again. I want something that allows me to go in random directions of my choosing.

**James Waldrop:**

Totally. I want the full power that the computer gives me, please.

**Beyang Liu:**

So, at Twitter, as manager of the developer productivity team, were you building additional tools internally for Twitter?

**James Waldrop:**

> Yeah. We became responsible pretty quickly, except for mobile, for the lifecycle. We thought of it as building testing and deploying.

Reviewer: ```suggestion
Yeah. We became responsible pretty quickly, except for mobile, for the lifecycle. We thought of it as building, testing, and deploying.
```

**Beyang Liu:**

Got it. And I think it was Review Board for code review and Pants for the build system?

**James Waldrop:**

I'm not at liberty to say so.

**Beyang Liu:**

Oh, sorry. Got it.

**James Waldrop:**

No, no, no, it's fine. Keep it in. We can talk about Pants. Twitter open sourced it. I was managing the team when we did that.

So, Twitter, as many Bay Area startups do, had a fair number of refugees from Google. Maybe more than most, I would say. And so, they had all used Blaze at Google. There was no Bazel at that point in time, and so they all missed their old build tool. So, Twitter, and it turns out Facebook did this too, wrote clones of Blaze and then Google went and open sourced Bazel. And so now we have three different Blaze-inspired open source build tools that people could use if they want to.

> But Twitter, of course, had theirs, Pants. Has theirs, still, I believe. I don't know, I haven't checked in to see if they've moved to anything else since I left. We were on Maven when I started. Maven was causing us a lot of trouble, a lot of trouble. It couldn't really walk the dependency graph in any reasonable way, and you really need that once you're in a mono repo. That's almost job one for your build tool on a mono repo. We had these crazy ass Maven POM XMLs that were deeply nested.

Reviewer: ```suggestion
But Twitter, of course, had theirs, Pants. Has theirs, still, I believe. I don't know, I haven't checked in to see if they've moved to anything else since I left. We were on Maven when I started. Maven was causing us a lot of trouble, a lot of trouble. It couldn't really walk the dependency graph in any reasonable way, and you really need that once you're in a monorepo. That's almost job one for your build tool on a monorepo. We had these crazy ass Maven POM XMLs that were deeply nested.
```

Yeah. Well, I still have it so I get my money's worth. At any rate, yeah, it was a hard problem. And so, Pants was our solution to that problem.

**Beyang Liu:**

Interesting. I guess taking a step back, I bet a good number of listeners are working on developer productivity, or are maybe part of a developer productivity team, or maybe even running it. What are your thoughts on how to run an effective dev prod organization? What are the things to look out for? What are the things that you measure? How do you define success?

**James Waldrop:**

It's good that you mentioned measurement. Measurement is a two-edged sword.

If you're on a team that has the goal to name itself “developer productivity,” you've got to be able to show that you're making people more productive, right?

**Beyang Liu:**

Yeah.

**James Waldrop:**

If you're not doing that, then you're not going to get headcount, that's for sure. Headcount is lifeblood for these teams. So, you need some metrics, but you really need to avoid being overly managed by your metrics. It can be very easy for those metrics to become targets that you have to hit as opposed to simply: are we improving life for developers?

> So, we measured things like build times, obviously. We knew all the slowest build targets. We could identify whether or not your machine needed to be upgraded because build targets were running slow for you, that we're running fast for other people. We could say, "Huh, there's something going on with your system. Let's figure out what it is." It's a corporate environment, so we could log all of the builds and ship them all up to our servers and see our metrics change as we release things. People loved that.

Reviewer: ```suggestion
So, we measured things like build times, obviously. We knew all the slowest build targets. We could identify whether or not your machine needed to be upgraded because build targets were running slow for you, that were running fast for other people. We could say, "Huh, there's something going on with your system. Let's figure out what it is." It's a corporate environment, so we could log all of the builds and ship them all up to our servers and see our metrics change as we release things. People loved that.
```

And so, I make everybody clear about what my motivations are. As long as I'm authentic about it, I feel like you can always find a way to align priorities. Of course, in the business, “Why?” was the very first question I asked at Fitbit because if it wasn't the priority, then I should go work on something else.

**Beyang Liu:**

Right.

**James Waldrop:**

But as long as you're working on the right set of priorities, then it's easy to get ... “Easy,” air quotes. It can be hard, but you will find a way to get people to support your effort, or whatever that is, and you don't need to be a manager.

Of course, I needed to be a manager to see that side of things before I realized how the management game works, and then go back to being an IC and use the knowledge I had gained going to the evil empire. Then I'm like, okay, I know how the evil empire works now so I can influence them from their perspective, from where they sit. And I can use language that works for them as well as for ICs.

> And so, understanding both perspectives I think is pretty important, at least it's been important for me. I think there's probably an 80% overlap, roughly, between what I would call the leadership circle of a Venn diagram. There is high-level IC and there's high level management, whatever that looks like and then there's leadership. There's a good amount of overlap between those two roles on a team or in a company.

Reviewer: ```suggestion
And so, understanding both perspectives I think is pretty important, at least it's been important for me. I think there's probably an 80% overlap, roughly, between what I would call the leadership circle of a Venn diagram. There is high-level IC and there's high-level management, whatever that looks like and then there's leadership. There's a good amount of overlap between those two roles on a team or in a company.
```
